superset/db_engine_specs/teradata.py

```diff
@@ -14,13 +14,223 @@
+        """
+        Return the table if valid, i.e., conforms to the [[catalog.]schema.]table
+        construct.
+
+        :param tlist: The SQL tokens
+        :returns: The table if the name conforms
+        """
+
+        # Strip the alias if present.
+        idx = len(tlist.tokens)
+
+        if tlist.has_alias():
+            ws_idx, _ = tlist.token_next_by(t=Whitespace)
+
+            if ws_idx != -1:
+                idx = ws_idx
+
+        tokens = tlist.tokens[:idx]
+
+        if (
+            len(tokens) in (1, 3, 5)
+            and all(imt(token, t=[Name, String]) for token in tokens[::2])
+            and all(imt(token, m=(Punctuation, ".")) for token in tokens[1::2])
+        ):
+            return Table(*[remove_quotes(token.value) for token in tokens[::-2]])
+
+        return None
+
+    @staticmethod
+    def _is_identifier(token: Token) -> bool:
+        return isinstance(token, (IdentifierList, Identifier))
+
+    def _process_tokenlist(self, token_list: TokenList) -> None:
+        """
+        Add table names to table set
+
+        :param token_list: TokenList to be processed
+        """
+        # exclude subselects
+        if "(" not in str(token_list):
+            table = self._get_table(token_list)
+            if table and not table.table.startswith(CTE_PREFIX):
+                self._tables.add(table)
+            return
+
+        # store aliases
+        if token_list.has_alias():
+            self._alias_names.add(token_list.get_alias())
+
+        # some aliases are not parsed properly
+        if token_list.tokens[0].ttype == Name:
+            self._alias_names.add(token_list.tokens[0].value)
+        self._extract_from_token(token_list)
+
+    def set_or_update_query_limit_td(self, new_limit: int) -> str:
+        td_sel_keywork = set(["SELECT ", "SEL "])
+        td_limit_keywork = set(["TOP", "SAMPLE"])
+        statement = self._parsed[0]
+
+        if not self._limit:
+            final_limit = new_limit
+        elif new_limit < self._limit:
+            final_limit = new_limit
+        else:
+            final_limit = self._limit
+
+        str_statement = str(statement)
+        str_statement = str_statement.replace("\n", " ").replace("\r", "")
+        tokens = str(str_statement).rstrip().split(" ")
+        tokens = list(filter(None, tokens))
+
+        next_remove_ind = False
+        new_tokens = []
+        for i in tokens:
+            if any(limitword in i.upper() for limitword in td_limit_keywork):
+                next_remove_ind = True
+            elif next_remove_ind and i.isdigit():
+                next_remove_ind = False
+            else:
+                new_tokens.append(i)
+                next_remove_ind = False
+
+        str_res = ""
+        for i in new_tokens:
+            str_res += i + " "
+            if any(selword in i.upper() for selword in td_sel_keywork):
+                str_res += "TOP " + str(final_limit) + " "
+        return str_res
```

Comment:
Traditionally in Python we build strings by appending to a list (the performance for adding strings used to be terrible, theses days it's equivalent):

```suggestion
        result = []
        for token in new_tokens:
            result.append(token)
            if any(selword in token.upper() for selword in td_sel_keywork):
                token.append(f"TOP {final_limit}")
        return " ".join(result)
```

(untested)